README.md
# n4dGtkLogincomponent
This is a Lliurex specific login component for Gtk apps.  
### API  
###### set_allowed_groups(groups=[])
	Login only will be succesfull if user belongs to any of this groups.
###### set_mw_proportion_ratio(columns_left,columns_right)  
	Sets the portion of screen reserved to each component (info and login form). It works splitting the main window in (columns_left+columns_right) columns and asigning to each part the desired number of columns. By default the proportion is set to 2:1
###### set_mw_background(image=None,cover=False,from_color=None,to_color=None,gradient='linear')  
	Sets the background for the login box. It can be a system's image or gradient going "from_color" to "to_color". By default all fields are "None" and only radial and linear gradients are supported. If we set a background image and set cover=True then the image will cover all the box area.
###### set_login_background(image=None,cover=False,from_color=None,to_color=None,gradient='linear')  
	Sets the background for the login box. It can be a system's image or gradient going "from_color" to "to_color". By default all fields are "None" and only radial and linear gradients are supported. If we set a background image and set cover=True then the image will cover all the box area.
###### set_default_username(default_username)  
	Sets the placeholder of the "username" entry to "default_username"  
###### set_default_server(default_server)  
	Sets the placeholder of the "server" entry to "default_server"  
###### set_login_banner(image)  
	Sets the user's image for the login form, by default is "llx-avatar" 
	If the image isn't a full path then is searched in the default theme.  
###### set_info_banner(image,resx=72,resy=72)  
	Sets the info box banner, by default is "None". The resolution is set to 72X72 by default.
###### set_info_background(image=None,cover=False,from_color=None,to_color=None,gradient='linear')  
	Sets the background for the info box. It can be a system's image or gradient going "from_color" to "to_color". By default all fields are "None" and only radial and linear gradients are supported. If we set a background image and set cover=True then the image will cover all the box area.
###### set_info_text(title,subtitle,text)  
	Sets the information to show in the info box.  
	It must have a title, a subtitle and a text as arguments and supports markup language.
###### hide_server_entry()  
	Hides the entry field of the login form
###### hide_info_box()  
	Hides the info box form
###### get_action_area()  
	Returns the info box so we can add any widget to it.  
###### after_validation_goto()  
	Sets the function that the loginComponent will launch after a correct user's validation. User, password and server will be passed as arguments to that function.
  
### Examples  
```python  
#!/usr/bin/env python3  
import gi  
gi.require_version('Gtk', '3.0')  
from gi.repository import Gtk  
from edupals.ui.n4dgtklogin import *  
  
def _signin():  
	print("OK")  
	print("Now hide the login component and make things")  
  
def start_gui():  
	mw=Gtk.Window()  
	loginComponent=N4dGtkLogin() #Init the login component  
#	loginComponent=N4dGtkLogin(orientation=Gtk.Orientation.VERTICAL) #Init the login component with vertical orientation
	loginComponent.set_info_text("<span foreground='black'>Title</span>","Subtitle","Text text text.\nText text text:\n<sub>* text with sub tag</sub>")  
	#Uncomment and comment...
	##Change the proportion ratio 
	#loginComponent.set_mw_proportion_ratio(2,1)
	#loginComponent.set_mw_proportion_ratio(1,2)
	#loginComponent.set_mw_proportion_ratio(3,9)
	#loginComponent.set_mw_proportion_ratio(8,5)
	##- Setting a background for the component
	#loginComponent.set_mw_background(image='/usr/share/backgrounds/ubuntu-mate-xenial/The_MATErix.png')  
	##- Setting a background for the form
	#loginComponent.set_form_background(from_color='grey',to_color='white',gradient='radial')  
	##- Adding a widget to the info box
	#infobox=loginComponent.get_action_area()  
	#infobox.add(Gtk.Label("Add widget"))  
	##- Changing the background for the info box
	#loginComponent.set_info_background(from_color='#000000',to_color='white',gradient='linear')  
	#loginComponent.set_info_background(image='/usr/share/backgrounds/ubuntu-mate-xenial/The_MATErix.png')  
	##- Changing default values for entries
	#loginComponent.set_default_username("Put your name")  
	#loginComponent.set_default_server("Put your server")  
	##- Changing banners
	#loginComponent.set_login_banner('/usr/share/filezilla/resources/flatzilla/48x48/uploadadd.png')  
	#loginComponent.set_info_banner('/usr/share/filezilla/resources/flatzilla/24x24/folder.png')  
	##- Function that will be launched after a succesfull validation
	loginComponent.after_validation_goto(_signin)  
	##
	mw.add(loginComponent)
	mw.connect("delete-event",Gtk.main_quit)  
	mw.show_all()  
  
start_gui()  
Gtk.main()  
```


debian/changelog
python-n4dgtklogin (1.1.3) bionic; urgency=medium

  * Fix python3 code on python2 version

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Mon, 09 Sep 2019 12:41:05 +0200

python-n4dgtklogin (1.1) bionic; urgency=medium

  * Another attempt... 

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Wed, 04 Sep 2019 13:37:30 +0200

python-n4dgtklogin (1.0.1) bionic; urgency=medium

  * Attempt to fix thread bug

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Mon, 29 Jul 2019 11:07:56 +0200

python-n4dgtklogin (1.0) bionic; urgency=medium

  * Add thread safety between glib and threads

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Wed, 24 Jul 2019 13:15:36 +0200

python-n4dgtklogin (0.6.5) bionic; urgency=medium

  * Launch thread into glib

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Fri, 05 Jul 2019 09:51:28 +0200

python-n4dgtklogin (0.6.4) bionic; urgency=medium

  * Fix default server on first boot

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Mon, 01 Jul 2019 12:51:10 +0200

python-n4dgtklogin (0.6.3) bionic; urgency=medium

  * Fix form fill on horizontal arrengement

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Mon, 01 Jul 2019 10:09:32 +0200

python-n4dgtklogin (0.6.2) bionic; urgency=medium

  * Fix set default server

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Fri, 28 Jun 2019 14:01:37 +0200

python-n4dgtklogin (0.6.1) bionic; urgency=medium

  * Change time interval for error message

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Thu, 27 Jun 2019 14:46:31 +0200

python-n4dgtklogin (0.6) bionic; urgency=medium

  * Add scrolling to main box
  * Replaced statusbar with animated revealer

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Fri, 21 Jun 2019 14:30:51 +0200

python-n4dgtklogin (0.5) bionic; urgency=medium

  * Check if localhost is also main server

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Fri, 01 Mar 2019 13:50:39 +0100

python-n4dgtklogin (0.4.7) bionic; urgency=medium

  * Update textfield icons

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Thu, 28 Feb 2019 14:03:28 +0100

python-n4dgtklogin (0.4.6.4) bionic; urgency=medium

  * Fix css for GtkEntry

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Fri, 23 Nov 2018 08:19:23 +0100

python-n4dgtklogin (0.4.6.3) bionic; urgency=medium

  * Fix spin position
  * Fix spinner bg

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Fri, 05 Oct 2018 10:22:40 +0200

python-n4dgtklogin (0.4.6.2) bionic; urgency=medium

  * Svn sync

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Tue, 02 Oct 2018 17:18:36 +0200

python-n4dgtklogin (0.4.5) bionic; urgency=medium

  * Bionic release

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Tue, 02 Oct 2018 17:01:41 +0200

python-n4dgtklogin (0.4.4) xenial; urgency=medium

  * Added vertical orientation 

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Thu, 21 Jun 2018 14:55:38 +0200

python-n4dgtklogin (0.4.3.1) xenial; urgency=medium

  * Added "server" to unknown host message

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Tue, 13 Mar 2018 14:34:44 +0100

python-n4dgtklogin (0.4.3) xenial; urgency=medium

  * Show server entry, although it has been hidden, when login can't connect to n4d at specified server

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Tue, 13 Mar 2018 14:13:13 +0100

python-n4dgtklogin (0.4.2) xenial; urgency=medium

  * Hide info label when there's no text to show

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Thu, 30 Nov 2017 09:05:41 +0100

python-n4dgtklogin (0.4.1) xenial; urgency=medium

  * Fix python2 init

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Tue, 28 Nov 2017 13:37:37 +0100

python-n4dgtklogin (0.4) xenial; urgency=medium

  * Change API 

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Mon, 27 Nov 2017 17:11:03 +0100

python-n4dgtklogin (0.3.3) xenial; urgency=medium

  * Use GLib for function call

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Thu, 23 Nov 2017 12:44:18 +0100

python-n4dgtklogin (0.3.2) xenial; urgency=medium

  * Add icon from theme as background

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Tue, 21 Nov 2017 13:58:29 +0100

python-n4dgtklogin (0.3.1) xenial; urgency=medium

  * add handler for socket exceptions

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Mon, 20 Nov 2017 10:09:33 +0100

python-n4dgtklogin (0.3) xenial; urgency=medium

  * Fix build

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Thu, 16 Nov 2017 14:34:20 +0100

python-n4dgtklogin (0.2-lliurex4) xenial; urgency=medium

  * Changed text domain to edupals

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Thu, 16 Nov 2017 12:26:36 +0100

python-n4dgtklogin (0.2-lliurex3) xenial; urgency=medium

  * Fix class init

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Wed, 15 Nov 2017 11:47:17 +0100

python-n4dgtklogin (0.2-lliurex2) xenial; urgency=medium

  * Added edupals-base as dependency

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Tue, 14 Nov 2017 14:46:34 +0100

python-n4dgtklogin (0.2-lliurex1) xenial; urgency=medium

  * Disabled pam

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Mon, 13 Nov 2017 16:42:40 +0100

python-n4dgtklogin (0.2) xenial; urgency=medium

  * Added python2 support

 -- lliurex <lliurex@localhost.localdomain>  Mon, 13 Nov 2017 08:48:43 +0100

python-n4dgtklogin (0.1) xenial; urgency=medium


setup3.py
#!/usr/bin/env python3
#
# $Id: setup.py,v 1.32 2010/10/17 15:47:21 ghantoos Exp $
#
#    Copyright (C) 2008-2009  Ignace Mouzannar (ghantoos) <ghantoos@ghantoos.org>
#
#    This file is part of lshell
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.

from setuptools import setup, find_packages
#from distutils.core import setup

if __name__ == '__main__':

    setup(name='n4dgtklogin',
        version='0.1',
        description='Python3 Gtk login component for N4D',
        long_description="""""",
        author='Lliurex Team',
        author_email='juanma1980@gmail.com',
        maintainer='Juanma Navarro',
        maintainer_email='juanma1980@gmail.com',
        keywords=['gtk','login','n4d'],
        url='http://www.lliurex.net',
        license='GPL',
        platforms='UNIX',
#        scripts = [''],
        package_dir = {'':'python3-n4dgtklogin.install'},
        packages = ['edupals.ui'],
        data_files = [],
        classifiers=[
                'Development Status :: 4 - Beta',
                'Environment :: Console'
                'Intended Audience :: End Users',
                'License :: OSI Approved :: GNU General Public License v3',
                'Operating System :: POSIX',
                'Programming Language :: Python',
                'Topic :: N4d login component',
                ],
    )



python-n4dgtklogin.install/edupals/ui/n4dgtklogin.py
#!/usr/bin/env python3
###
#This class returns a login_grid whith a standarized login form
###
import os,sys,socket
import threading
import netifaces 
import gi
gi.require_version('Gtk', '3.0')
gi.require_version('PangoCairo', '1.0')
import json
import cairo
#import commands
from gi.repository import Gtk, Gdk, GdkPixbuf, GObject, GLib, PangoCairo, Pango
try:
	import xmlrpclib as n4d
except ImportError:
	raise ImportError("xmlrpc not available. Disabling server queries")
import time
import gettext

gettext.textdomain('edupals.ui.common')
_ = gettext.gettext
GObject.threads_init()

class N4dGtkLogin(Gtk.Box):
	__gtype_name__='n4dgtklogin'

	def __init__(self,*args,**kwds):
		super(Gtk.Box,self).__init__(*args,**kwds)
		self.dbg=True
		self.msgCount=0
		self.vertical=False
		if 'orientation' in kwds.keys():
			if kwds['orientation']==Gtk.Orientation.VERTICAL:
				self.vertical=True
		self.sw_n4d=True
		if hasattr(sys,'last_value'):
		#If there's any error at this point it only could be an ImportError caused by xmlrpc
			self.sw_n4d=False
		self.css_classes={'#GtkEntry':'{font-family: Roboto;border:0px;border-bottom:1px grey solid;margin-top:0px;padding-top:0px;}'}
		self.style_provider=Gtk.CssProvider()
		Gtk.StyleContext.add_provider_for_screen(Gdk.Screen.get_default(),self.style_provider,Gtk.STYLE_PROVIDER_PRIORITY_APPLICATION)
		self.default_spacing=12
		self.username_placeholder=_("Username")
		self.server_placeholder=_("Server IP (Default value : server)")
		self.default_server='server'
		self.login_banner_default="llx-avatar"
		self.info_banner=None
		self.allowed_groups=[]
		#internal boxes
		self.form_box=Gtk.Box(spacing=0,orientation=Gtk.Orientation.VERTICAL)
		self.info_box=Gtk.Box(spacing=self.default_spacing,orientation=Gtk.Orientation.VERTICAL)
		self.main_grid=Gtk.Grid()
		self.render_form()
	#def __init__

	def _debug(self,msg):
		if self.dbg:
			print("%s",msg)

	def set_allowed_groups(self,groups):
		self.allowed_groups=groups
	#def set_allowed_groups

	def set_mw_proportion_ratio(self,left_panel,right_panel):
		for child in self.main_grid.get_children():
			self.main_grid.remove(child)
		if self.vertical:
			self.main_grid.attach(self.form_box,0,0,1,left_panel)
			self.main_grid.attach(self.info_box,0,0+left_panel,1,right_panel)
			self.main_grid.set_margin_bottom(12)
		else:
			self.main_grid.attach(self.info_box,0,0,left_panel,1)
			self.main_grid.attach(self.form_box,0+left_panel,0,right_panel,1)
	#def set_mw_proportion_ratio
	
	def set_mw_background(self,image=None,cover=False,from_color='#ffffff',to_color='silver',gradient='linear'):
		mw_background=self._set_background(image,cover,from_color,to_color,gradient)
		self.css_classes['#mw']='{'+mw_background+';;}'
		self._set_css()
	#def set_mw_background

	def set_login_background(self,image=None,cover=False,from_color='#ffffff',to_color='@silver',gradient='linear'):
		form_background=self._set_background(image,cover,from_color,to_color,gradient)
		self.css_classes['#main']='{'+form_background+';;}'
		self._set_css()
	#def set_login_background

	def set_info_background(self,image=None,cover=False,from_color='#ffffff',to_color='@silver',gradient='linear'):
		info_background=self._set_background(image,cover,from_color,to_color,gradient)
		self.css_classes['#info']='{'+info_background+';;}'
		self._set_css()
	#def set_info_background

	def _set_css(self):
		css=''
		for css_class,style in self.css_classes.items():
			css=css+css_class+' '+style
		css_style=eval('b"""'+css+'"""')
		self.style_provider.load_from_data(css_style)
	#def _set_css

	def _set_background(self,image=None,cover=False,from_color='#ffffff',to_color='silver',gradient='linear'):
		bg=''
		if image and os.path.isfile(image):
			if cover:
				bg='background-image:url("'+image+'"); background-repeat:no-repeat; background-size:100% 100%'
			else:
				bg='background-image:url("'+image+'"); background-repeat:no-repeat;'
		elif image:
			self._debug("%s not found. Searching..."%image)
			#try to locate the image in the default theme
			icon_theme=Gtk.IconTheme.get_default()
			icon_sizes=icon_theme.get_icon_sizes(image)
			if icon_sizes:
				max_size=max(icon_sizes)
				icon=icon_theme.lookup_icon(image,max_size,0)
				icon_path=icon.get_filename()
				bg='background-image:url("'+icon_path+'"); background-repeat:no-repeat; background-size:100% 100%'

		else:
			if gradient=='linear':
				bg='background-image:-gtk-gradient (linear, left top, left bottom, from ('+from_color+'),  to ('+to_color+'))'
			elif gradient=='radial':
				bg='background-image:-gtk-gradient (radial, center center,0,center center,1, from ('+from_color+'),  to ('+to_color+'))'
		return bg
	#def _set_background

	def set_default_username(self,username):
		self.username_placeholder=username
		self._set_text_for_entry(self.txt_username,username)
	#def set_default_username

	def set_default_server(self,server):
		self.server_placeholder=server
		self.default_server=server
		if self.txt_server:
			self._set_text_for_entry(self.txt_server,server)
	#def set_default_server
	
	def _set_text_for_entry(self,widget,text):
		widget.set_placeholder_text(text)
	#def _set_text_for_entry

	def _lookup_user_face(self):
		sw_ok=False
		if os.path.isfile(os.path.expanduser('~/.face')):
			sw_ok=True
			self.set_login_banner(os.path.expanduser('~/.face'))
		return sw_ok
	#def _lookup_user_face

	def set_login_banner(self,image=None):
		avatar=None
		if image:
			avatar=self._get_image(image)
			if avatar==None:
				avatar=self._get_image("avatar-default")
			self.login_banner.set_from_pixbuf(avatar)
	#def set_banner

	def set_info_banner(self,banner,x=72,y=72):
		self.info_banner.set_from_pixbuf(self._get_image(banner))
	#def set_info_banner

	def _get_image(self,image,x=72,y=72):
		pixbuf=None
		icon_theme=Gtk.IconTheme.get_default()
		if icon_theme.has_icon(image):
			pixbuf=icon_theme.load_icon(image,x,0)
		else:
			if os.path.isfile(image):
				pixbuf=GdkPixbuf.Pixbuf.new_from_file_at_scale(image,x,y,True)
		return pixbuf
	#def _get_image
	
	def set_info_text(self,title,subtitle,text):
		sw_ok=True
		info_msg=''
		self.lbl_info_msg.set_width_chars(25)
		self.lbl_info_msg.set_max_width_chars(25)
		try:
			msg="<b><big>"+title+"</big></b>"
			info_msg=msg+'\n'+subtitle+'\n'+text
		except Exception as e:
			sw_ok=False
			print(e)
		self.lbl_info_msg.set_markup(info_msg)
		self.lbl_info_msg.show()
		return sw_ok
	#def set_info_text

	def hide_server_entry(self):
		self.txt_server.props.no_show_all=True
		self.txt_server.hide()
		return True
	#def hide_server_entry

	def show_info(self,msg):
		if self.msgCount<10:
			self.msgCount+=1
		elif self.rev_info.get_reveal_child():
			self.rev_info.set_transition_type(Gtk.RevealerTransitionType.SLIDE_UP)
			self.rev_info.set_reveal_child(False)
			self.msgCount=0
			return False
		self.rev_info.set_transition_type(Gtk.RevealerTransitionType.SLIDE_DOWN)
		self.rev_info.set_reveal_child(True)
		return True
    #def show_info

	def hide_info_box(self):
		self.info_box.props.no_show_all=True
		self.info_box.hide()
	#def hide_info_box

	def get_action_area(self):
		return self.info_box
	#def get_action_area

	def render_form(self):
		self.scr=Gtk.ScrolledWindow()
		self.scr.set_policy(Gtk.PolicyType.AUTOMATIC,Gtk.PolicyType.AUTOMATIC)
		self.scr.set_propagate_natural_height(True)
		self.scr.set_propagate_natural_width(True)
		sb=self.scr.get_vscrollbar()
		sb.hide()
		self.main_grid.set_hexpand(True)
		self.main_grid.set_vexpand(True)

		self.main_grid.set_column_homogeneous(True)
		self.main_grid.set_row_homogeneous(True)
		self._render_login_form()
		self._render_info_form()
		if self.vertical:
			self.main_grid.attach(self.form_box,1,1,1,1)
			self.main_grid.attach(self.info_box,1,3,1,1)
			self.main_grid.set_margin_bottom(12)
			self.scr.set_valign(Gtk.Align.START)
			self.pack_start(self.scr,False,False,0)
		else:
			self.main_grid.attach(self.info_box,1,1,2,1)
			self.main_grid.attach(self.form_box,3,1,1,1)
			self.pack_start(self.scr,True,True,0)
		self.scr.add(self.main_grid)
		self.set_name("mw")
		self.form_box.set_name("main")
	#def render_form

	def _render_login_form(self):
		self.txt_username=Gtk.Entry()
		self.txt_username.set_name("GtkEntry")
		self.login_banner=Gtk.Image()
		if not self._lookup_user_face():
			self.set_login_banner(self.login_banner_default)
		self.login_banner.set_margin_bottom(self.default_spacing)
		self.set_default_username(self.username_placeholder)
		self.txt_username.set_icon_from_icon_name(Gtk.EntryIconPosition.PRIMARY,"system-users")
		self.txt_password=Gtk.Entry()
		self.txt_password.set_name("GtkEntry")
		self.txt_password.set_icon_from_icon_name(Gtk.EntryIconPosition.PRIMARY,"changes-prevent")
		self._set_text_for_entry(self.txt_password,_("Password"))
		self.txt_server=Gtk.Entry()
		self.txt_server.set_name("GtkEntry")
#		self.set_default_server(self.server_placeholder)
		self.txt_server.set_icon_from_icon_name(Gtk.EntryIconPosition.PRIMARY,"network-wired")
		self.btn_sign=Gtk.Button(stock=Gtk.STOCK_OK)
		self.btn_sign.connect('clicked',self._validate)
		self.btn_sign.set_halign(Gtk.Align.CENTER)
		login_grid=Gtk.Grid()
		login_grid.set_column_spacing(6)
		login_grid.set_row_spacing(6)
		login_grid.set_margin_left(self.default_spacing)
		login_grid.set_margin_right(self.default_spacing)
		login_grid.set_margin_top(self.default_spacing)
		login_grid.set_margin_bottom(self.default_spacing)
		self.spinner=Gtk.Spinner()
		self.spinner.set_name("spinner")
		if self.vertical:
			position=Gtk.PositionType.RIGHT
		else:
			position=Gtk.PositionType.BOTTOM

		login_grid.attach(self.spinner,0,1,2,5)
		login_grid.attach(self.txt_username,0,1,1,1)
		self._set_widget_default_props(self.txt_username,_("Username"))
		self.txt_username.connect('activate',self._validate)
		login_grid.attach_next_to(self.txt_password,self.txt_username,position,1,1)
		self._set_widget_default_props(self.txt_password,_("Password"))
		self.txt_password.set_visibility(False)
		self.txt_password.props.caps_lock_warning=True
		self.txt_password.connect('activate',self._validate)
		self._set_widget_default_props(self.txt_server,_("Master Server IP"))
		if self.vertical:
			self.txt_server.set_halign(Gtk.Align.FILL)
			self.txt_server.set_hexpand(True)
			login_grid.attach_next_to(self.txt_server,self.txt_username,Gtk.PositionType.BOTTOM,2,1)
		else:
			login_grid.attach_next_to(self.txt_server,self.txt_password,Gtk.PositionType.BOTTOM,1,1)
		self.txt_server.connect('activate',self._validate)
		self.btn_sign.set_margin_top(self.default_spacing)
		login_grid.attach_next_to(self.btn_sign,self.txt_server,Gtk.PositionType.BOTTOM,3,1)
		login_grid.attach(self.login_banner,0,0,3,1)
		self.rev_info=Gtk.Revealer()
		self.rev_info.set_transition_duration(1000)
		self.rev_info.set_valign(Gtk.Align.START)
		self.rev_info.set_name("revealer")
		if not '#revealer' in self.css_classes.keys():
				self.css_classes['#revealer']='{background-color:red;color:white;padding:3px;}'
		self.lbl_error=Gtk.Label(_("Login failed"))
		self.lbl_error.set_name("revealer")
		self.rev_info.add(self.lbl_error)
		login_grid.attach(self.rev_info,0,0,3,1)
		if self.vertical:
			login_grid.props.valign=Gtk.Align.END
			login_grid.set_margin_bottom(0)
			self.btn_sign.set_margin_bottom(0)
		else:
			login_grid.props.valign=Gtk.Align.CENTER
		login_grid.props.halign=Gtk.Align.CENTER
		self.form_box.pack_start(login_grid,True,True,0)
	#def _render_login_form

	def _render_info_form(self):
		self.info_box.set_homogeneous(False)
		self.info_banner=Gtk.Image()
		info_detail_box=Gtk.Box(spacing=self.default_spacing,orientation=Gtk.Orientation.VERTICAL)
		info_detail_box.pack_start(self.info_banner,False,False,0)
		self.lbl_info_msg=Gtk.Label()
		self.lbl_info_msg.set_use_markup(True)
		self.lbl_info_msg.set_line_wrap(True)
		info_detail_box.pack_start(self.lbl_info_msg,True,True,0)
		if not '#label' in self.css_classes.keys():
			self.css_classes['#label']='{background-color:rgba(200,200,200,0.8);;}'
		self.lbl_info_msg.set_name("label")
		self.lbl_info_msg.set_no_show_all(True)
		self.lbl_info_msg.hide()
		self.info_box.set_name("info")
		if self.vertical:
			self.lbl_info_msg.set_justify(Gtk.Justification.CENTER)
			self.lbl_info_msg.set_margin_left(12)
			self.lbl_info_msg.set_margin_right(12)
			info_detail_box.props.valign=Gtk.Align.START
		else:
			info_detail_box.props.valign=Gtk.Align.CENTER
			info_detail_box.props.halign=Gtk.Align.CENTER
		if self.vertical:
			self.info_box.pack_start(info_detail_box,False,False,0)
		else:
			self.info_box.pack_start(info_detail_box,True,True,0)
		self._set_css()
	#def _render_info_form

	def set_label_background(self,r,g,b,a):
		self.css_classes['#label']='{background-color:rgba(%s,%s,%s,%s);;}'%(r,g,b,a)

	def _set_widget_default_props(self,widget,tooltip=None):
		widget.set_valign(True)
		widget.set_halign(True)
		widget.set_tooltip_text(tooltip)
	#def _set_widget_default_props

	def _status_info_hide(self,*args):
		self.rev_info.hide()
	#def _info_hide

	def _validate(self,widget=None):
		height=self.get_allocation().height

		self.scr.set_min_content_height(height)
		self.scr.set_propagate_natural_height(False)
		self.scr.set_propagate_natural_width(False)
		self.scr.set_max_content_height(height+1)
		user=self.txt_username.get_text()
		pwd=self.txt_password.get_text()
		server=self.txt_server.get_text()
		server_ip=''
		if not server:
			server=self.default_server
		try:
			server_ip=socket.gethostbyname(server)
		except:
			server='localhost'
		#Check if localhost is server
		for iface in netifaces.interfaces():
			for key,netinfo in netifaces.ifaddresses(iface).items():
				for info in netinfo:
					if info['addr']==server_ip:
						server='localhost'
						break

		self.form_box.set_sensitive(False)
		self.spinner.start()
		self.th=threading.Thread(target=self._t_validate,args=[user,pwd,server])
		self.th.start()
	#def _validate
	
	def _t_validate(self,user,pwd,server):
		Gdk.threads_enter()
		ret=[False]
		msg=_("Login failed")
		if self.sw_n4d:
			try:
				self.n4dclient=self._n4d_connect(server)
				ret=self.n4dclient.validate_user(user,pwd)
			except socket.error as e:
				msg=_("Unknown host")
				ret=[False,str(e)]

		if (isinstance(ret,bool)):
			pass
		elif not ret[0]:
			pass
		elif self.allowed_groups and not set(self.allowed_groups).intersection(ret[1]):
			#Check user groups
			msg=_("User not allowed")
		else:
			msg=""
		GLib.idle_add(self._validate_result,msg,user,pwd,server)
		Gdk.threads_leave()
		return(False)
	#def _t_validate
	
	def _validate_result(self,msg,user,pwd,server):
		self.spinner.stop()
		self.form_box.set_sensitive(True)
		self.th.join()
		if msg:
			GLib.timeout_add(500,self.show_info,(msg))
			self.txt_server.props.no_show_all=False
			self.txt_server.show()
		else:
			self.after_validate(user,pwd,server)
		return False

	def after_validation_goto(self,func,data=None):
		self.after_validate=func
	#def after_validation_func

	def _n4d_connect(self,server):
		try:
			socket.gethostbyname(server)
		except:
			#It could be an ip
			try:
				socket.inet_aton(server)
			except Exception as e:
				print(e)
				raise
		c = n4d.ServerProxy("https://"+server+":9779")
		return c
	#def _n4d_connect
